src/Functions/FunctionStringCompare.cpp

```diff
@@ -0,0 +1,414 @@
+
+    String getName() const override { return name; }
+    size_t getNumberOfArguments() const override { return 0; }
+    bool isVariadic() const override { return true; }
+    bool isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo & /*arguments*/) const override { return false; }
+
+    DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
+    {
+        if (arguments.size() != 2 && arguments.size() != 5)
+            throw Exception(ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH, "Function {} requires 2 or 5 arguments", getName());
+
+        FunctionArgumentDescriptors mandatory_args{
+            {"string1", static_cast<FunctionArgumentDescriptor::TypeValidator>(&isStringOrFixedString), nullptr, "String"},
+            {"string2", static_cast<FunctionArgumentDescriptor::TypeValidator>(&isStringOrFixedString), nullptr, "String"},
+        };
+        FunctionArgumentDescriptors optional_args{
+            {"string1_offset", static_cast<FunctionArgumentDescriptor::TypeValidator>(&isUInt), static_cast<FunctionArgumentDescriptor::ColumnValidator>(&FunctionStringCompare::isConstColumn), "const UInt*"},
+            {"string2_offset", static_cast<FunctionArgumentDescriptor::TypeValidator>(&isUInt), static_cast<FunctionArgumentDescriptor::ColumnValidator>(&FunctionStringCompare::isConstColumn), "const UInt*"},
+            {"num_bytes", static_cast<FunctionArgumentDescriptor::TypeValidator>(&isUInt), static_cast<FunctionArgumentDescriptor::ColumnValidator>(&FunctionStringCompare::isConstColumn), "const UInt*"},
+        };
+        validateFunctionArguments(*this, arguments, mandatory_args, optional_args);
+
+        return std::make_shared<DataTypeInt8>();
+    }
+
+    DataTypePtr getReturnTypeForDefaultImplementationForDynamic() const override { return std::make_shared<DataTypeInt8>(); }
+
+    ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr & /*result_type*/, size_t input_rows_count) const override
+    {
+        chassert(arguments[0].column->size() == input_rows_count);
+        auto col_result = ColumnInt8::create();
+        if (!input_rows_count)
+            return std::move(col_result);
+
+        auto prepare_string_args
+            = [](const ColumnPtr & column) -> std::tuple<const ColumnString *, const ColumnFixedString *, const ColumnConst *>
+        {
+            return {
+                checkAndGetColumn<ColumnString>(column.get()),
+                checkAndGetColumn<ColumnFixedString>(column.get()),
+                checkAndGetColumn<ColumnConst>(column.get())};
+        };
+        auto [col_str1, fixed_str1_col, col_str1_const] = prepare_string_args(arguments[0].column);
+        auto [col_str2, fixed_str2_col, col_str2_const] = prepare_string_args(arguments[1].column);
+
+        auto str1_offset = arguments.size() > 2 ? arguments[2].column->getUInt(0) : 0;
```

Comment:
Can we define `bool has_five_args = arguments.size() > 5` and use it in l. 82, 83, 85?